documentation/puppet-api/v3/environment_classes.md

```diff
@@ -0,0 +1,385 @@
+
+The environment must be purely alphanumeric, not ''
+```
+
+#### Environment does not include only alphanumeric characters
+
+A request made with an environment parameter which includes at least one
+character which is not among the following - A-Z, a-z, 0-9, or _ (underscore) -
+results in an HTTP 400 (Bad Request) error:
+
+```
+GET /puppet/v3/environment_classes?environment=bog|us
+
+HTTP/1.1 400 Bad Request
+
+The environment must be purely alphanumeric, not 'bog|us'
+```
+
+### Schema
+
+An environment classes response body conforms to
+[the environment classes schema](./environment_classes.json).
+ 
+### Headers and Caching Behavior
+
+If the `environment-class-cache-enabled` setting in the `jruby-puppet`
+configuration section is set to `true`, the `environment_classes` API can
+make use of caching for the response data.  This can provide a significant
+performance benefit and reduction in the amount of data that needs to be
+provided in a response when the underlying Puppet code on disk remains
+unchanged from one request to the next.  Use of the cache does, however,
+require that cache entries are invalidated after Puppet code has been
+updated.  To avoid having to invalidate cache entries, the
+`environment-class-cache-enabled` setting can be omitted from configuration 
+or explicitly set to `false`.  In this case, manifests will be re-discovered
+from disk and re-parsed for every incoming request.  This can involve
+significantly greater overhead in performance and bandwidth for repeated 
+requests where the underlying Puppet code does not change much.  This 
+approach does, however, ensure that the latest available data is returned 
+for every request made.  The rest of the content in this section pertains to the
+behaviors to keep in mind when the cache setting is enabled.
+
+When the `environment-class-cache-enabled` setting is set to `true`, the 
+response to a query to the `environment_classes`  endpoint includes an HTTP
+[Etag](https://tools.ietf.org/html/rfc7232#section-2.3) header.  The value 
+for the Etag header is a hash which represents the state of the latest class
+information available for the requested environment.  For example:
+
+```
+ETag: 31d64b8038258202b4f5eb508d7dab79c46327bb
+```
+
+A client may (but is not required to) provide the Etag value back to the server
+in a subsequent `environment_classes` request.  The client would provide the
+tag value as the value for an
+[If-None-Match](https://tools.ietf.org/html/rfc7232#section-3.2)
+HTTP header:
+
+```
+If-None-Match: 31d64b8038258202b4f5eb508d7dab79c46327bb
+```
+
+If the latest state of code available on the server matches that of the value
+in the `If-None-Match` header, the server will provide an HTTP 304 (Not 
+Modified) response and no corresponding response body.  If the server has later
+code available than what is captured by the `If-None-Match` header value, the
+server will re-parse code from manifest files on disk.  Assuming the resulting
```

Comment:
I assume it will also do this if no `If-None-Match` header is given, or if an invalid header is included in the request?